File: 0098-Validate Binary Search Tree/0098-Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if (root == nullptr)
        {
            return true;
        }
        
        stack<TreeNode*> St;
        TreeNode* node = root;
        TreeNode* prev = nullptr;
        while (!St.empty() || node)
        {
            if (node)
            {
                St.push(node);
                node = node->left;
            }
            else
            {
                node = St.top();
                if (prev && node->val <= prev->val)
                {
                    return false;
                }
                prev = node;
                St.pop();
                node = node->right;
            }
        }
        
        return true;
    }
};

File: 0001-0100/0098-Validate Binary Search Tree/0098-Validate Binary Search Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        curr = root
        St = []
        prev = None
        while St or curr:
            if curr:
                St.append(curr)
                curr = curr.left
            else:
                curr = St.pop()
                if prev and prev.val >= curr.val:
                    return False
                prev = curr
                curr = curr.right
        return True
